ci/run.sh

```diff
@@ -94,61 +101,77 @@ run_one () {
     BUILDTYPE=$3
 
     TAG="shadow:${CONTAINER/:/-}-$CC-$BUILDTYPE"
-    echo "Running $TAG"
-
-    # Build and tag a Docker image with the given configuration.
-    docker build -t "$TAG" $NOCACHE -f- . <<EOF
-    FROM $CONTAINER
-
-    ENV CONTAINER $CONTAINER
-    SHELL ["/bin/bash", "-c"]
-
-    # Install base dependencies before adding CC or BUILTYPE to
-    # the environment, allowing this layer to be cached and reused
-    # across other images for this CONTAINER.
-    COPY ci/container_scripts/install_deps.sh /root/install_deps.sh
-    RUN /root/install_deps.sh
-
-    # Now install any CC/BUILDTYPE-specific dependencies.
-    ENV CC=$CC BUILDTYPE=$BUILDTYPE
-    COPY ci/container_scripts/install_extra_deps.sh /root/install_extra_deps.sh
-    RUN /root/install_extra_deps.sh
-    ENV PATH /root/.cargo/bin:\$PATH
-
-    # Copy the local source into the container.
-    RUN mkdir /root/shadow
-    COPY . /root/shadow
-    WORKDIR /root/shadow
-
-    # Build and install. We do this as part of building the Docker image to support
-    # reusing such images for incremental builds (TBD). 
-    RUN ci/container_scripts/build_and_install.sh
+
+    if [ "${BUILD_IMAGES}" == "1" ]; then
+        echo "Building $TAG"
+
+        # Build and tag a Docker image with the given configuration.
+        docker build -t "$TAG" $NOCACHE -f- . <<EOF
+        FROM $CONTAINER
+
+        ENV CONTAINER $CONTAINER
+        SHELL ["/bin/bash", "-c"]
+
+        # Install base dependencies before adding CC or BUILTYPE to
+        # the environment, allowing this layer to be cached and reused
+        # across other images for this CONTAINER.
+        COPY ci/container_scripts/install_deps.sh /root/install_deps.sh
+        RUN /root/install_deps.sh
+
+        # Now install any CC/BUILDTYPE-specific dependencies.
+        ENV CC=$CC BUILDTYPE=$BUILDTYPE
+        COPY ci/container_scripts/install_extra_deps.sh /root/install_extra_deps.sh
+        RUN /root/install_extra_deps.sh
+        ENV PATH /root/.cargo/bin:\$PATH
+
+        # Copy the local source into the container.
+        RUN mkdir /root/shadow
+        COPY . /root/shadow
+        WORKDIR /root/shadow
+
+        # Build and install. We do this as part of building the Docker image to support
+        # reusing such images for incremental builds (TBD).
+        RUN ci/container_scripts/build_and_install.sh
 EOF
-    echo "Testing $TAG"
+    fi
 
     # Run the tests inside the image we just built.
-    docker run --rm --shm-size=1g $TAG /bin/bash ci/container_scripts/test.sh
+    echo "Testing $TAG"
+
+    # Start the container
+    CONTAINER_ID="$(docker create --shm-size=1g --cap-add=SYS_PTRACE "$TAG" /bin/bash -c \
+        "ci/container_scripts/build_and_install.sh && ci/container_scripts/test.sh")"
+
+    # Copy the most recent code into the container
+    docker cp . "${CONTAINER_ID}":/root/shadow
```

Comment:
I played around with this PR a little bit locally. I seem to be wrong about the mtimes. It appears to be rebuilding a few object files every time for some reason, but not most of them. We also save some time with the lighter-weight docker operations and not having to re-download the cargo deps.

We definitely need to somehow avoid copying the local build directory into the container, though.

Reply:
> I played around with this PR a little bit locally. I seem to be wrong about the mtimes. It appears to be rebuilding a few object files every time for some reason, but not most of them. We also save some time with the lighter-weight docker operations and not having to re-download the cargo deps.

It's not rebuilding any object files for me. What version of Docker are you using, and is your host OS using something other than ext4?

> We definitely need to somehow avoid copying the local build directory into the container, though.

I didn't consider this. We can just:

```
docker cp -a . "${CONTAINER_ID}":/root/shadow-tmp
```

and later in the container:

```
rm -r /root/shadow-tmp/{.git,build,src/main/tmp}
cp -ar /root/shadow-tmp .
```

Is there an advantage to doing the volume/rsync approach in this case? But we should figure out why it's rebuilding some of your files first.

Reply:
I think the only correctness issue with that is that it'll leave deleted files in place. That might lead to confusing behavior if a source file is removed or renamed but is still referenced by the build system or includes, etc. i.e. instead of breaking the build as it ought to, it'll just use the stale file.

It may also be a little slower since it copies the build and .git directories before removing them, and it copies source files that weren't modified. The latter probably doesn't matter much, but we probably could and should fix the former by doing something like `D=`mktemp -d` && find . -maxdepth 1 -not -name .git -not -name build -exec cp -ar \{\} $D \;`

That still leaves the correctness issue of not removing files that no longer exist locally. It *probably* doesn't matter that much, but OTOH is there a reason not to go the volume + rsync route instead?

Reply:
Seems to be working with rsync. Down-side is that rsync would be another dependency to install. Not that big of a deal, but if you can work out a way to do it without rsync that works for me too :)

```
+        RUN apt-get install -y rsync
+        RUN mkdir /mnt/shadow
+        VOLUME /mnt/shadow
+
         # Copy the local source into the container.
         RUN mkdir /root/shadow
         COPY . /root/shadow
@@ -139,11 +143,11 @@ EOF
     echo "Testing $TAG"
 
     # Start the container
-    CONTAINER_ID="$(docker create --shm-size=1g --cap-add=SYS_PTRACE "$TAG" /bin/bash -c \
-        "ci/container_scripts/build_and_install.sh && ci/container_scripts/test.sh")"
+    CONTAINER_ID="$(docker create --shm-size=1g --cap-add=SYS_PTRACE --mount src=$(realpath .),target=/mnt/shadow,type=bind,readonly "$TAG" /bin/bash -c \
+        "rsync --del --exclude=build --exclude=.git -r /mnt/shadow/  . && ci/container_scripts/build_and_install.sh && ci/container_scripts/test.sh")"
 
     # Copy the most recent code into the container
-    docker cp . "${CONTAINER_ID}":/root/shadow
+    #docker cp . "${CONTAINER_ID}":/root/shadow
```

Reply:
I wrote an option that doesn't use rsync, but it's not worth the additional complexity when rsync is only a few hundred KB. I'll add an rsync version later today.